blackjax/adaptation/chees_adaptation.py

```diff
@@ -0,0 +1,426 @@
+            acceptance_probabilities,
+            is_divergent,
+            adaptation_state,
+        )
+
+        return ChEESAdaptationState(*parameters)
+
+    return init, update
+
+
+def chees_adaptation(
+    logprob_fn: Callable,
+    num_chains: int,
+    *,
+    jitter_generator: Optional[Callable] = None,
+) -> AdaptationAlgorithm:
+    """Adapt the step size and trajectory length (number of integration steps / step size)
+    parameters of the jittered HMC algorthm.
+
+    The jittered HMC algorithm depends on the value of a step size, controlling
+    the discretization step of the integrator, and a trajectory length, given by the
+    number of integration steps / step size, jittered by using only a random percentage
+    of this trajectory length.
+
+    This adaptation algorithm tunes the trajectory length by heuristically maximizing
+    the Change in the Estimator of the Expected Square (ChEES) criterion over
+    an ensamble of parallel chains. At equilibrium, the algorithm aims at eliminating
+    correlations between target dimensions, making the HMC algorithm efficient.
+
+    Jittering requires generating a random sequence of uniform variables in [0, 1].
+    However, this adds another source of variance to the sampling procedure,
+    which may slow adaptation or lead to suboptimal mixing. To alleviate this,
+    rather than use uniform random noise to jitter the trajectory lengths, we use a
+    quasi-random Halton sequence, which ensures a more even distribution of trajectory
+    lengths.
+
+    Examples
+    --------
+
+    An HMC adapted kernel can be learned and used with the following code:
+
+    .. code::
+
+        warmup = blackjax.chees_adaptation(logprob_fn, num_chains)
+        key_warmup, key_sample = jax.random.split(rng_key)
+        optim = optax.adam(learning_rate)
+        (last_states, parameters), _ = warmup.run(
+            key_warmup,
+            positions, #PyTree where each leaf has shape (num_chains, ...)
+            initial_step_size,
+            optim,
+            num_warmup_steps,
+        )
+        kernel = blackjax.dynamic_hmc(logprob_fn, **parameters).step
+        new_states, info = jax.vmap(kernel)(key_sample, last_states)
+
+    Parameters
+    ----------
+    logprob_fn
+        The log density probability density function from which we wish to sample.
+    num_chains
+        Number of chains used for cross-chain warm-up training.
+    jitter_generator
+        Optional function that generates a value in [0, 1] used to jitter the trajectory
+        lengths given a PRNGKey, used to propose the number of integration steps. If None,
+        then a quasi-random Halton is used to jitter the trajectory length.
+
+    Returns
+    -------
+    A function that returns the last cross-chain state, a sampling kernel with the
+    tuned parameter values, and all the warm-up states for diagnostics.
+
+    """
+
+    def run(
+        rng_key: PRNGKey,
+        positions: ArrayLikeTree,
```

Comment:
given this is top level API, worth adding an assertion to make sure positions have leading axis size = num_chains